crates/oxc_transformer/src/plugins/styled_components.rs

```diff
@@ -0,0 +1,1154 @@
+                        }
+                        _ => return None,
+                    };
+                }
+                // `const X = styled`
+                Ancestor::VariableDeclaratorInit(declarator) => {
+                    return if let BindingPatternKind::BindingIdentifier(ident) =
+                        &declarator.id().kind
+                    {
+                        Some(ident.name)
+                    } else {
+                        None
+                    };
+                }
+                // `const X = { Y: styled }`
+                Ancestor::ObjectPropertyValue(property) => {
+                    return if let PropertyKey::StaticIdentifier(ident) = property.key() {
+                        Some(ident.name)
+                    } else {
+                        None
+                    };
+                }
+                // `class Y { (static) X = styled }`
+                Ancestor::PropertyDefinitionValue(property) => {
+                    return if let PropertyKey::StaticIdentifier(ident) = property.key() {
+                        Some(ident.name)
+                    } else {
+                        None
+                    };
+                }
+                _ => {
+                    if ancestor.is_parent_of_statement() {
+                        // we've hit a statement, we should stop crawling up
+                        return assignment_name;
+                    }
+                }
+            }
+        }
+
+        unreachable!()
+    }
+
+    fn get_component_id(&mut self, ctx: &TraverseCtx<'a>) -> &'a str {
+        let namespace = self.options.namespace.as_ref().map_or("", |namespace| {
+            ctx.ast.allocator.alloc_concat_strs_array([namespace.as_str(), "__"])
+        });
+
+        let file_hash = self.get_file_hash(ctx);
+        let id = ctx.ast.allocator.alloc_concat_strs_array([
+            namespace,
+            "sc-",
+            file_hash,
+            "-",
+            itoa::Buffer::new().format(self.component_count),
+        ]);
+        self.component_count += 1;
+        id
+    }
+
+    fn get_file_hash(&mut self, ctx: &TraverseCtx<'a>) -> &'a str {
+        use rustc_hash::FxHasher;
+        use std::hash::{Hash, Hasher};
+
+        #[inline]
+        fn base36_encode<'a>(mut num: u64, ctx: &TraverseCtx<'a>) -> &'a str {
+            const BASE36_BYTES: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
+
+            num %= 36_u64.pow(6); // 36^6, to ensure the result is <= 6 characters long.
+
+            let mut result = Vec::with_capacity(6);
+
+            while num != 0 {
+                result.push(BASE36_BYTES[(num % 36) as usize]);
+                num /= 36;
+            }
+
+            ctx.ast.allocator.alloc_str(
+                // SAFETY: the bytes are valid UTF-8 as they are ASCII characters.
+                unsafe { std::str::from_utf8_unchecked(&result) },
+            )
+        }
```

Comment:
Is there a reason for using base 36 here? Couldn't we use base 64 (A-Z, a-z, 0-9, _, $) instead?

Building the string with `InlineString` would be be more performant than `Vec`. Like in mangler:

https://github.com/oxc-project/oxc/blob/fe843806ece40a19adca3a7c6366abfb6f8662bb/crates/oxc_mangler/src/base54.rs

Reply:
I guess we can use base64. Using `base36` is to align with its Babel plugin.

https://github.com/styled-components/babel-plugin-styled-components/blob/4e2eb388d9c90f2921c306c760657d059d01a518/src/utils/hash.js#L54-L55. 

Ok, I will follow this kind of optimization.

Reply:
It's not a big thing! Was just asking if there was a reason because base36 is unusual.

Reply:
@Dunqing I realized now that maybe there *is* a reason. Maybe class names are case insensitive in some browsers?